polymorphia-frontend/services/api/client.ts

```diff
@@ -0,0 +1,87 @@
+import { API_HOST } from "@/services/api";
+import { BackendErrorResponse } from "@/interfaces/api/error";
+import { ApiError } from "@/services/api/error";
+import handleUnauthorized from "@/services/api/handle-unauthorized";
+import { ApiBody, ApiRequestOptions } from "@/services/api/types";
+
+const GENERIC_ERROR_MESSAGE = "Wystąpił nieoczekiwany błąd. Spróbuj ponownie.";
+
+const readErrorMessage = async (response: Response): Promise<string> => {
+  try {
+    const { detail, title } = (await response.json()) as BackendErrorResponse;
+    return detail?.trim() || title?.trim() || GENERIC_ERROR_MESSAGE;
+  } catch {
+    return GENERIC_ERROR_MESSAGE;
+  }
+};
+
+const buildUrl = (path: string): string =>
+  `${API_HOST}${path.startsWith("/") ? path : `/${path}`}`;
+
+const buildHeaders = (
+  headers: HeadersInit | undefined,
+  hasJsonBody: boolean
+): Headers | undefined => {
+  if (!headers && !hasJsonBody) {
+    return undefined;
+  }
+  const result = new Headers(headers ?? {});
```

Comment:
```suggestion
  const result = new Headers(headers);
```
With the condition above it's not necessary.